src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
                 return false;
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            if (!EncodeString(value, values, separator, buffer.Slice(i), out int valueLength, lowercase: false))
             {
                 return false;
             }
-
             i += valueLength;
 
             length = i;
             return true;
         }
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int i, bool lowercase)
         {
             const int toLowerMask = 0x20;
+            // TODO: use huffman encoding
+            for (int j = 0; j < s.Length; j++)
+            {
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
 
+                //buffer[i++] = (byte)(s[j] | (lowercase && s[j] >= 'A' && s[j] <= 'Z' ? toLowerMask : 0));
+                buffer[i++] = (byte)(s[j] | (lowercase ? toLowerMask : 0));
+            }
```

Comment:
Let's add a TODO to use https://github.com/dotnet/corefx/issues/34144 when it's available.